infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/DefaultHostIngressExternalServerExposer.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.server;
+
+import static java.lang.Integer.parseInt;
+import static java.util.stream.Collectors.toMap;
+
+import io.fabric8.kubernetes.api.model.ServicePort;
+import io.fabric8.kubernetes.api.model.extensions.Ingress;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.config.ServerConfig;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides a path-based strategy for exposing service ports outside the cluster using Ingress
+ * Ingresses will be created without an explicit host (defaulting to *).
+ *
+ * <p>This strategy uses different Ingress path entries <br>
+ * Each external server is exposed with a unique path prefix.
+ *
+ * <p>This strategy imposes limitation on user-developed applications. <br>
+ * It should only be used for local development with a single IP address
+ *
+ * <pre>
+ *   Path-Based Ingress exposing service's port:
+ * Ingress
+ * ...
+ * spec:
+ *   rules:
+ *     - http:
+ *         paths:
+ *           - path: service123/webapp        ---->> Service.metadata.name + / + Service.spec.ports[0].name
+ *             backend:
+ *               serviceName: service123      ---->> Service.metadata.name
+ *               servicePort: [8080|web-app]  ---->> Service.spec.ports[0].[port|name]
+ * </pre>
+ *
+ * @author Sergii Leshchenko
+ * @author Guy Daich
+ */
+public class DefaultHostIngressExternalServerExposer
+    implements ExternalServerExposerStrategy<KubernetesEnvironment> {
+
+  public static final String DEFAULT_HOST_STRATEGY = "default-host";
+  private final Map<String, String> ingressAnnotations;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(DefaultHostIngressExternalServerExposer.class);
+
+  @Inject
+  public DefaultHostIngressExternalServerExposer(
+      @Named("infra.kubernetes.ingress.annotations") Map<String, String> ingressAnnotations) {
+    if (ingressAnnotations == null) {
+      LOG.warn(
+          "Ingresses annotations are absent. Make sure that workspace ingresses don't need "
+              + "to be configured according to ingress controller.");
+    }
+    this.ingressAnnotations = ingressAnnotations;
+  }
+
+  @Override
+  public void exposeExternalServers(
```

Comment:
Looks like three strategies classes have very similar code with ingress builders. Maybe it would be better (maybe not) to create abstract `IngressExternalServerExposerStrategy` with a protected method like `#createIngress(String name, @Nullable String host, String path, ...)`

Reply:
created AbstractIngressExternalServerExposerStrategy, where shared logic resides.